src/Source/Webapi.php

```diff
@@ -0,0 +1,103 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Jh\Import\Source;
+
+use Countable;
+use Exception;
+use Jh\Import\Report\Report;
+use Jh\Import\Source\Webapi\CountResponseHandlerInterface;
+use Jh\Import\Source\Webapi\DataRequest\FilterDecoratorInterface;
+use Jh\Import\Source\Webapi\DataRequest\PagingDecoratorInterface;
+use Jh\Import\Source\Webapi\DataResponseHandlerInterface;
+use Jh\Import\Source\Webapi\RequestFactoryInterface;
+use Psr\Http\Client\ClientInterface;
+
+class Webapi implements Source, Countable
+{
+    private ?int $totalNumberOfItems = null;
+    private string $idField;
+    private string $sourceId;
+    private RequestFactoryInterface $countRequestFactory;
+    private CountResponseHandlerInterface $countResponseHandler;
+    private RequestFactoryInterface $dataRequestFactory;
+    private int $dataRequestPageSize;
+    private PagingDecoratorInterface $dataRequestPagingDecorator;
+    private DataResponseHandlerInterface $dataResponseHandler;
+    private ClientInterface $httpClient;
+    private ?FilterDecoratorInterface $dataRequestFilterDecorator;
+
+    public function __construct(
+        string $idField,
+        string $sourceId,
+        RequestFactoryInterface $countRequestFactory,
+        CountResponseHandlerInterface $countResponseHandler,
+        RequestFactoryInterface $dataRequestFactory,
+        int $dataRequestPageSize,
+        PagingDecoratorInterface $dataRequestPagingDecorator,
+        DataResponseHandlerInterface $dataResponseHandler,
+        ClientInterface $httpClient,
+        ?FilterDecoratorInterface $dataRequestFilterDecorator = null
+    ) {
+        $this->idField = $idField;
+        $this->sourceId = $sourceId;
+        $this->countRequestFactory = $countRequestFactory;
+        $this->countResponseHandler = $countResponseHandler;
+        $this->dataRequestFactory = $dataRequestFactory;
+        $this->dataRequestPagingDecorator = $dataRequestPagingDecorator;
+        $this->dataRequestPageSize = $dataRequestPageSize;
+        $this->dataResponseHandler = $dataResponseHandler;
+        $this->httpClient = $httpClient;
+        $this->dataRequestFilterDecorator = $dataRequestFilterDecorator;
+    }
+
+    public function count(): int
+    {
+        if (!$this->totalNumberOfItems) {
```

Comment:
One to think about - maybe add the support of null here?
The thing is, if the count response gives you back `0` as a response, you will keep calling that request all the time, even though that request has already been made. My point being - maybe make this field nullable and only do the API call if it is null, otherwise return the value (even if it's `0`)?

Reply:
Ok fixed, I have added additional check before actual migration starts, as 0 count collapses progress bar.